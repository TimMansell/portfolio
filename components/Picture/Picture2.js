export const Picture = () => {
  return <>test</>;
};

export default Picture;
