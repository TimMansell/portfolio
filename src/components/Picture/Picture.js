import React from 'react';
import PropTypes from 'prop-types';

import styles from './Picture.module.scss';

export const Picture = ({ srcs, name, defaultImg }) => {
  const fallbackImg = defaultImg || srcs[0].src;

  return (
    <picture title={name} data-e2e="picture">
      {srcs.map(({ src, format }, index) => (
        <source
          type={`image/${format}`}
          srcSet={src}
          key={index}
          data-test={`picture-source-${index}`}
        />
      ))}

      <img
        className={styles.img}
        alt={name}
        src={fallbackImg}
        data-test="picture-img"
      />
    </picture>
  );
};

Picture.propTypes = {
  srcs: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string.isRequired,
      format: PropTypes.string.isRequired,
    })
  ).isRequired,
  name: PropTypes.string.isRequired,
  defaultImg: PropTypes.string,
};

export default Picture;
