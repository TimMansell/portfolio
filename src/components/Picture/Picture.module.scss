.img {
  max-width: 100%;
  height: auto;
  object-fit: cover;
}

.fullscreen {
  height: 100%;
}
