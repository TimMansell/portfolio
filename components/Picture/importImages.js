import { getSources, getFallbackImage } from './helpers';

export const importImages = (src, srcSizes, types) => {
  const sources = types.flatMap(getSources(src, srcSizes));
  const fallbackImg = getFallbackImage(src, srcSizes, types);

  return {
    sources,
    fallbackImg,
  };
};
