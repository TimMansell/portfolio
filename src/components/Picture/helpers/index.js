import { getSizes } from './importImages';

export const getSources = (src, srcSizes) => (type) =>
  srcSizes.map(getSizes(src, type));

export const getFallbackImage = (src, srcSizes, types) => {
  const [fallbackSrc] = [...srcSizes].reverse();
  const { sizes } = fallbackSrc;
  const [fallbackSize] = [...sizes].reverse();
  const [fallbackType] = [...types].reverse();

  const fallbackImg = require(`../../${src}-${fallbackSize}.${fallbackType}`);

  return fallbackImg;
};
