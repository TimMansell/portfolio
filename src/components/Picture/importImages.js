import { getSources, getFallbackImage } from './helpers';

export const importImages = (src, types, srcSizes) => {
  const sources = types.map(getSources(src, srcSizes));
  const fallbackImg = getFallbackImage(src, srcSizes, types);

  // console.log(JSON.stringify(sources));

  return {
    sources: sources[0],
    fallbackImg,
  };
};
