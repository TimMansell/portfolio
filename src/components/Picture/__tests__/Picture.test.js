import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Picture from '../Picture';

const props = {
  srcs: [
    {
      src: 'test src',
      format: 'jpg',
    },
  ],
  name: 'test name',
};

describe('Picture', () => {
  it('should render my component', () => {
    // eslint-disable-next-line
    const wrapper = shallow(<Picture {...props} />);
  });

  it('should match snapshot', () => {
    const snapshot = renderer.create(<Picture {...props} />).toJSON();

    expect(snapshot).toMatchSnapshot();
  });

  it('should have correct <source> properties', () => {
    const wrapper = shallow(<Picture {...props} />);
    const source = wrapper.find('[data-test="picture-source-0"]');

    expect(source.props().srcSet).toBe(props.srcs[0].src);
    expect(source.props().type).toBe(`image/${props.srcs[0].format}`);
  });

  it('should render srcs[0] in <img>', () => {
    const wrapper = shallow(<Picture {...props} />);
    const source = wrapper.find('[data-test="picture-img"]');

    expect(source.props().src).toBe(props.srcs[0].src);
    expect(source.props().alt).toBe(props.name);
  });

  it('should render defaultImg in <img>', () => {
    const defaultImageProps = {
      ...props,
      defaultImg: 'default src',
    };
    const wrapper = shallow(<Picture {...defaultImageProps} />);
    const source = wrapper.find('[data-test="picture-img"]');

    expect(source.props().src).toBe(defaultImageProps.defaultImg);
    expect(source.props().alt).toBe(props.name);
  });
});
