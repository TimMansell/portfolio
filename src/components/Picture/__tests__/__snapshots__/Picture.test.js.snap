// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picture should match snapshot 1`] = `
<picture
  data-e2e="picture"
  title="test title"
>
  <source
    data-test="picture-source-0"
    srcSet="test.avif"
    type="image/avif"
  />
  <source
    data-test="picture-source-1"
    srcSet="test.webp"
    type="image/webp"
  />
  <img
    alt="test title"
    className="img"
    data-e2e="picture-img"
    data-test="picture-img"
    loading="auto"
    src="test.webp"
  />
</picture>
`;
